dali/operators/audio/preemphasis_filter_op.cu

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/audio/preemphasis_filter_op.h"
+#include <vector>
+
+namespace dali {
+
+namespace detail {
+
+template <typename OutputType, typename InputType>
+struct SampleDescriptor {
+  const InputType *in;
+  OutputType *out;
+  float coeff;
+  int64_t size;
+};
+
+template <typename OutputType, typename InputType>
+void __global__ PreemphasisFilterKernel(const SampleDescriptor<OutputType, InputType> *samples) {
+  const auto &sample = samples[blockIdx.y];
+  int64_t block_size = blockDim.x;
+  int64_t block_start = block_size * blockIdx.x;
+  int64_t grid_stride = block_size * gridDim.x;
+  for (int64_t k = block_start + threadIdx.x; k < sample.size; k += grid_stride) {
+    auto prev = cuda_max(k-1, 0l);
+    sample.out[k] = sample.in[k] - sample.coeff * sample.in[prev];
+  }
+}
+
+}  // namespace detail
+
+template <>
+void PreemphasisFilter<GPUBackend>::RunImpl(workspace_t<GPUBackend> &ws) {
+  const auto &input = ws.InputRef<GPUBackend>(0);
+  auto &output = ws.OutputRef<GPUBackend>(0);
+  TYPE_SWITCH(input.type().id(), type2id, InputType, PREEMPH_TYPES, (
+    TYPE_SWITCH(output_type_, type2id, OutputType, PREEMPH_TYPES, (
+      using SampleDesc = detail::SampleDescriptor<OutputType, InputType>;
+
+      std::vector<SampleDesc> samples_cpu(batch_size_);
+      for (int sample_idx = 0; sample_idx < batch_size_; sample_idx++) {
+        auto &sample = samples_cpu[sample_idx];
+        sample.in = input.tensor<InputType>(sample_idx);
+        sample.out = output.mutable_tensor<OutputType>(sample_idx);
+        sample.size = volume(input.tensor_shape(sample_idx));
+        sample.coeff = preemph_coeff_[sample_idx];
+      }
+
+      scratchpad_.set_type(TypeInfo::Create<uint8_t>());
+      int64_t sz = batch_size_ * sizeof(SampleDesc);
+      scratchpad_.Resize({sz});
+      auto sample_descs_gpu = reinterpret_cast<SampleDesc*>(scratchpad_.mutable_data<uint8_t>());
+      auto stream = ws.stream();
+      CUDA_CALL(
+        cudaMemcpyAsync(sample_descs_gpu, samples_cpu.data(), sz, cudaMemcpyHostToDevice, stream));
+
+      int block = 1024;
```

Comment:
My recent experience suggests, that blocks this large are not necessarily optimal. If you happen to have a benchmark, please check 256, 384 and 512.

Reply:
I'll benchmark and choose the best